bitcoin/BRWalletManager.c

```diff
@@ -1191,3 +1130,234 @@ bwmAnnounceSubmit (BRWalletManager manager,
+
+                // store the first unused addresses for comparison in the next complete call
+                bwm->brdSync.lastExternalAddress = externalAddress;
+                bwm->brdSync.lastInternalAddress = internalAddress;
+
+                // get the addresses to query the BDB with
+                BRWalletManagerGetAllAddrsAsStrings (bwm,
+                                                    &addressCount,
+                                                    &addressStrings,
+                                                    &addressArray);
+
+                // don't need to alter the range (we haven't found all transactions yet)
+
+                // store sync data for callback outside of lock
+                begBlockNumber = bwm->brdSync.chunkBegBlockNumber;
+                endBlockNumber = bwm->brdSync.chunkEndBlockNumber;
+
+            } else if (bwm->brdSync.chunkEndBlockNumber != bwm->brdSync.endBlockNumber) {
+                // .. we haven't discovered any new addresses but we haven't gone through the whole range yet
+
+                // don't need to store the first unused addresses (we just confirmed they are equal)
+
+                // get the addresses to query the BDB with
+                BRWalletManagerGetAllAddrsAsStrings (bwm,
+                                                    &addressCount,
+                                                    &addressStrings,
+                                                    &addressArray);
+
+                // store the new range
+                bwm->brdSync.chunkBegBlockNumber = bwm->brdSync.chunkEndBlockNumber;
+                bwm->brdSync.chunkEndBlockNumber = MIN (bwm->brdSync.chunkEndBlockNumber + BWM_BRD_SYNC_CHUNK_SIZE,
+                                                        bwm->brdSync.endBlockNumber);
+
+                // store sync data for callback outside of lock
+                begBlockNumber = bwm->brdSync.chunkBegBlockNumber;
+                endBlockNumber = bwm->brdSync.chunkEndBlockNumber;
+
+            } else {
+                // .. we haven't discovered any new addresses and we just finished the last chunk
+
+                // reset sync state and advance the sync range by updating `begBlockNumber`
+                bwm->brdSync.completed           = 1;
+                bwm->brdSync.lastInternalAddress = BR_ADDRESS_NONE;
+                bwm->brdSync.lastExternalAddress = BR_ADDRESS_NONE;
+                bwm->brdSync.begBlockNumber      = (bwm->brdSync.endBlockNumber >=  BWM_BRD_SYNC_START_BLOCK_OFFSET
+                                                    ? bwm->brdSync.endBlockNumber - BWM_BRD_SYNC_START_BLOCK_OFFSET
+                                                    : 0);
+            }
+        } else {
+            // reset sync state on failure
+            bwm->brdSync.lastInternalAddress = BR_ADDRESS_NONE;
+            bwm->brdSync.lastExternalAddress = BR_ADDRESS_NONE;
+            bwm->brdSync.completed           = 1;
+        }
```

Comment:
bwmSyncReset() is not appropriate here (in the above `else`)?

Reply:
My personal style/preference is to not call out for this type of case. I prefer to have this function "self-contained" (you could make the argument that the state is reset above too and could be a call to bwmSyncReset() as well).

The reset call has locking and chunk implications that I'd like to avoid as well.

Reply:
okay.